src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -67,16 +68,34 @@ private BoundExpression MakeIsConstantPattern(BoundConstantPattern loweredPatter
 
         private BoundExpression MakeIsDeclarationPattern(BoundDeclarationPattern loweredPattern, BoundExpression loweredInput)
         {
-            Debug.Assert(loweredPattern.Variable.GetTypeOrReturnType() == loweredPattern.DeclaredType.Type);
+            Debug.Assert(((object)loweredPattern.Variable == null && loweredPattern.VariableAccess.Kind == BoundKind.DiscardedExpression) ||
+                         loweredPattern.Variable.GetTypeOrReturnType() == loweredPattern.DeclaredType.Type);
 
             if (loweredPattern.IsVar)
             {
-                Debug.Assert(loweredInput.Type == loweredPattern.Variable.GetTypeOrReturnType());
-                var assignment = _factory.AssignmentExpression(loweredPattern.VariableAccess, loweredInput);
                 var result = _factory.Literal(true);
+
+                if (loweredPattern.VariableAccess.Kind == BoundKind.DiscardedExpression)
+                {
+                    return result;
+                }
+
+                Debug.Assert((object)loweredPattern.Variable != null && loweredInput.Type == loweredPattern.Variable.GetTypeOrReturnType());
+
+                var assignment = _factory.AssignmentExpression(loweredPattern.VariableAccess, loweredInput);
                 return _factory.MakeSequence(assignment, result);
             }
 
+            if (loweredPattern.VariableAccess.Kind == BoundKind.DiscardedExpression)
+            {
+                var temps = ArrayBuilder<LocalSymbol>.GetInstance(1);
+                BoundLocal discard = MakeTempForDiscardedExpression((BoundDiscardedExpression)loweredPattern.VariableAccess, temps);
```

Comment:
>MakeTempForDiscardedExpression [](start = 37, length = 30)

Rather than allocating a builder for a single local, consider adding a method that simply gives the symbol back (split MakeTempForDiscardedExpression in two methods).